aip/0233.md

```diff
@@ -0,0 +1,82 @@
+---
+aip:
+  id: 233
+  state: draft
+  created: 2019-06-18
+permalink: /233
+redirect_from:
+  - /0233
+---
+
+# Batch Create
+
+Some APIs need to allow users to create multiple resources in a single
+transaction. A batch create method provides this functionality.
+
+## Guidance
+
+Batch create methods are specified using the following pattern:
+
+```proto
+rpc BatchCreateBooks(BatchCreateBooksRequest) returns (BatchCreateBooksResponse) {
+  option (google.api.http) = {
+    post: "/v1/shelves/-/books:batchCreate"
+    body: "*"
+  };
+}
+```
+
+- The RPC's name **must** begin with `BatchCreate`. The remainder of
+  the RPC name **should** be the plural form of the resource being created.
+- The request and response messages **must** match the RPC name, with `-Request`
+  and `-Response` suffixes.
+  - However, in the event that the request may take a significant amount of
+    time, the response message **must** be a `google.longrunning.Operation`
+    which ultimately resolves to the `-Response` type.
+- The HTTP verb **must** be `POST`.
+- The HTTP URI **must** end with `:batchCreate`.
+- The URL path **should** represent the collection for the resource, matching
+  the collection used for simple CRUD operations. If the operation spans
+  parents, a dash (`-`) **may** be used as a wildcard.
```

Comment:
I am a bit conflicted about wildcards here. If we require the parent, then that means a single IAM check. Otherwise, the service has to go through and do all the IAM checks before starting any actual work, and that sounds like something that would be easy for services to mess up.

Perhaps "allow but discourage" this?

Reply:
Yeah, I'm leaning towards keep Batch** calls limited within a parent. It seems like otherwise, atomicity isn't so critical if one is spanning parents.

Reply:
I don't know - I think it's possible that you might want to create two resources that refer to each other, but have different parents. I'm conflicted too, basically.

Reply:
I discussed this with @jgeewax today and we believe that including the parent should be the common case, but we can allow APIs to break the glass by supporting `-` if they have a good reason.